### PR TITLE
[ot_certs] ASN1 codegen

### PR DESCRIPTION
**Pull request by pamaury**
Only review the last three commits, others are from #20601. See commits for description.
To test:
```bash
# build all files
./bazelisk.sh build //sw/device/silicon_creator/lib/cert/...
# view all targets:
./bazelisk.sh query //sw/device/silicon_creator/lib/cert/...
# query list of output for a certificate:
./bazelisk.sh outquery-all //sw/device/silicon_creator/lib/cert:example_template
# use your favorite editor to look at those files
```

### CODE REVIEW

File: sw/host/ot_certs/src/codegen.rs [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+  int32_t arg = 0;
+  char mod_id[5];
+  bool is_err = status_extract(err, &code, &arg, mod_id);
+  mod_id[4] = 0;
+  if(is_err) {
+    fprintf(stderr, "An error occured: %s[%d,%s]\n", code, arg, mod_id);
+  }
+  return 0;
+}
+"#;
```
Comment (jwnrt) [inline sw/host/ot_certs/src/codegen.rs:70]: Nit / suggestion: you may find it clearer to put this in a `.c` file in this directory and use `include_str!` to pull it in.

File: rules/certificates.bzl [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    )
+    return [
+        DefaultInfo(files = depset(outputs)),
+        OutputGroupInfo(
+            sources = depset([out_c]),
+            headers = depset([out_h]),
+        ),
+    ]
+
+# TODO: document
```
Comment (jwnrt) [inline rules/certificates.bzl:41]: Would you mind adding these doc strings before we merge?

It could easily be forgotten if we leave it to the future

**Review by jwnrt — APPROVED** [review]
I've done a pass on this and it looks generally okay, but there's really lots of generation code and I definitely haven't checked its correctness in depth.

I don't have any refactorings to suggest, but if you have anything that could make this easier to follow or to make it more compact, that could be helpful for the future.

**Comment by pamaury** [comment]
Thanks for the comments James, I will rework that. Regarding the code generation, I agree that it's a lot of code but I don't have any great idea to make the code more compact. I'll think about it when updating the code.

**Comment by pamaury** [comment]
I have removed the code of the writer from this PR since it takes quite a bit of code and it is not essential. I have addressed some changes, I still need to fix two TODO.

File: sw/device/silicon_creator/lib/cert/BUILD [hunk 3]
```diff
@@ -2,6 +2,36 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
+load(
+    "//rules:certificates.bzl",
+    "certificate_template",
+)
+
+certificate_template(
+    "generic_template",
+    "generic.hjson",
```
Comment (timothytrippel) [inline sw/device/silicon_creator/lib/cert/BUILD:12]:
```suggestion
    name = "generic_template",
    template = "generic.hjson",
```

applies throughout

File: rules/certificates.bzl [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        OutputGroupInfo(
+            sources = depset([out_c]),
+            headers = depset([out_h]),
+        ),
+    ]
+
+# This rule uses `opentitantool certificate codegen` to generate the
+# source files of the C certificate generator. Those source files
+# need to be compiled and linked with the asn1 library: this is NOT
+# done by this rule. The generates files are put in the following
```
Comment (timothytrippel) [inline rules/certificates.bzl:44]:
```suggestion
# done by this rule. The generated files are put in the following
```

File: rules/certificates.bzl [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        arguments = [
+            "--rcfile=",
+            "--quiet",
+            "certificate",
+            "codegen",
+            "--template={}".format(ctx.file.template.path),
+            "--output-c={}".format(out_c.path),
+            "--output-h={}".format(out_h.path),
+        ],
+        executable = tc.tools.opentitantool,
```
Comment (timothytrippel) [inline rules/certificates.bzl:30]: we probably also want to hit the generated C files with clang format after they are generated?

**Review by timothytrippel — APPROVED** [review]

File: sw/host/ot_certs/src/asn1/codegen.rs [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+        str_type: &Tag,
+        val: &Value<String>,
+    ) -> Result<()> {
+        let str_type = str_type.codestring();
+        match val {
+            Value::Literal(x) => {
+                // NOTE: we are not using a constant for the size of the string because technically there is no way
+                // to know what the `char` type is and therefore what is the length of this string for the C compiler.
+                self.push_str_with_indent(&format!("RETURN_IF_ERROR(asn1_push_string(&state, {str_type}, \"{x}\", strlen(\"{x}\")));\n"));
+                // A tagged string needs a tag, up to 3 bytes of length and the string itself.
```
Comment (jwnrt) [inline sw/host/ot_certs/src/asn1/codegen.rs:419]: Nit: can we put `(up to three bytes)` in parens?

File: sw/host/ot_certs/src/asn1/codegen.rs [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+                // Don't try to exactly compute how many bytes we need for the length.
+                self.max_out_size += 1 + 3 + x.len();
```
Comment (jwnrt) [inline sw/host/ot_certs/src/asn1/codegen.rs:421]: Nit: if the `1` is for the null terminator, can we put it at the end to make that clear?

File: sw/host/ot_certs/src/codegen.rs [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    let mut source_h = String::new();
+
+    let license_and_warning = indoc::formatdoc! { r#"
+        // Copyright lowRISC contributors.
+        // Licensed under the Apache License, Version 2.0, see LICENSE for details.
+        // SPDX-License-Identifier: Apache-2.0
+
+        // This file was automatically generated using opentitantool from:
+        // {from_file}
+        "#};
```
Comment (jwnrt) [inline sw/host/ot_certs/src/codegen.rs:41]: Nit: de-indent this one level

**Review by jwnrt — APPROVED** [review]
Reviewed this with @luismarques and we only had these nits

File: rules/certificates.bzl (hunk 5 above)
Comment (pamaury) [reply rules/certificates.bzl:30]: I will look into it.

**Comment by pamaury** [comment]
I have pushed an update that rebases the code and also addresses the various comments. I have also improved the documentation and the code now produces better estimates for the size of the certificate (previous estimates were in the ball park of 550 bytes, now it is at 440 so it's much better!).

File: sw/host/ot_certs/src/codegen.rs [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+
+    // This file was automatically generated using opentitantool from:
+    // {from_file}
+    "#};
+
+    // License, warning about autogenerated code and guard inclusion checks.
+    source_c.push_str(&license_and_warning);
+    source_h.push_str(&license_and_warning);
+    let preproc_guard_include = tmpl.name.to_uppercase();
+    source_h.push_str(&format!("#ifndef __{}__\n", preproc_guard_include));
```
Comment (nbdd0121) [inline sw/host/ot_certs/src/codegen.rs:70]:
```suggestion
    write!(source_h, "#ifndef __{preproc_guard_include}__\n");
```

**Pull request merged**